crypto/rand/drbg_lib.c

```diff
@@ -90,6 +90,19 @@ static RAND_DRBG *drbg_private;
  * |randomness| argument). This will immediately reseed the <master> DRBG.
  * The <public> and <private> DRBG will detect this on their next generate
  * call and reseed, pulling randomness from <master>.
+ *
+ * LOCKING
+ *
+ * The three shared DRBGs are intended to be used concurrently, so they
+ * support locking by default. It is the callers responsibility to wrap
+ * calls to functions like RAND_DRBG_generate() which modify the DRBGs
+ * internal state with calls to RAND_DRBG_lock() and RAND_DRBG_unlock().
```

Comment:
Why is it the caller's responsibility?

Reply:
This simply documents the status quo: currently the RAND_METHOD callbacks already take the lock whenever the shared RAND_DRBGs (master, public, and private) are accessed.

Initially, I intended to move the locking logic entirely into the RAND_DRBG, but for good reason I stepped back from doing this. A short explanation is in https://github.com/openssl/openssl/issues/4403#issuecomment-364375168.  TL;DR: Since locking is optional and currently enabled for the three shared drbgs, managing the locking logic internally only adds unecessary complexity. If you are interested in a more detailed explanation, it will take a little time to write it down.

Reply:
To make it clear: locking is _not_ required if you simply use `RAND_bytes()` or `RAND_priv_bytes()`, because these functions manage the locking internally. It's only needed for direct acces to the `RAND_DRBG` api.

Reply:
I think I need to dig deeper into the DRBG code to grasp it fully, but ok.

Reply:
... and only for DRBGs that are accessed concurrently, in particular not for the SSL DRBG instances.

Reply:
Ahhhh, that last comment of yours made sense to me.  Ok then!

Reply:
> I think I need to dig deeper into the DRBG code to grasp it fully, but ok.

You don’t have to dig deep into the DRBG code if you are busy with other things. I’ll be content if @kroeckx and maybe @kaduk looks over it.

Reply:
Perhaps add a "(such as already done by RAND_bytes)" to the comment.

Reply:
Added the following lines

```
 * The functions RAND_bytes() and RAND_priv_bytes() take the locks
 * automatically, so using the RAND api is thread safe as before.
```